### PR TITLE
Rewrite `dolt status` subcommand to use SQL queries to generate results.

### PR DESCRIPTION
**Pull request by PavelSafronov**
Rewrite `dolt status` subcommand to use SQL queries to generate results.

This change adds `remote` column to the `dolt_branches` table. This column identifies the remote for the branch.
This change adds `dolt_count_commits` function, which returns the number of commits between two commits.
Added more BATS tests for `dolt status`.

### CODE REVIEW

File: go/cmd/dolt/dolt.go
Comment (macneale4) [inline go/cmd/dolt/dolt.go:564]: Talk to nick about this. I think he has something better on the way

File: go/cmd/dolt/commands/status.go
Comment (macneale4) [inline go/cmd/dolt/commands/status.go:132]: There are a lot of separate calls to the server for this. Ideas on how we can reduce the number of round trips?

**Review by macneale4 — CHANGES_REQUESTED** [review]
Not sure if all commands are going to require this many round trips. Let's discuss before shipping. Asked for one additional bats test too

File: go/cmd/dolt/dolt.go
Comment (PavelSafronov) [reply go/cmd/dolt/dolt.go:564]: Nick's PR has been approved. Once that's merged, I'll rebase my changes and this should no longer be necessary.

File: go/cmd/dolt/commands/status.go
Comment (PavelSafronov) [reply go/cmd/dolt/commands/status.go:129]: Removed one call that was unnecessary.
From offline convo, it was decided that there can be improvements in this area, but those can wait until after we ship these features.

File: integration-tests/bats/status.bats
Comment (PavelSafronov) [reply integration-tests/bats/status.bats:460]: Adding.

File: integration-tests/bats/sql-local-remote.bats
Comment (macneale4) [inline integration-tests/bats/sql-local-remote.bats:128]: You should drop the --verbose-engine-setup flag since you aren't looking for the output from it (and I don't think you need to). Update three over uses below.

**Review by macneale4 — APPROVED** [review]

**Comment by macneale4** [comment]
Looks like there are failures in `bats/constraint-violations.bats` immediately after calling `dolt status` - something isn't right.

File: go/cmd/dolt/dolt.go
Comment (macneale4) [reply go/cmd/dolt/dolt.go:564]: Rebase onto main. You'll get a conflict, but Nick's new ParseGlobalArgs method is better than my hack

File: integration-tests/bats/sql-local-remote.bats
Comment (macneale4) [inline integration-tests/bats/sql-local-remote.bats:129]: I think a test which would be easier to read and more correct would be to take all of the output from the remote invocation, and compare it with all the output of the local invocation. The point of the test is to verify that the two modes produce the same output, not to make sure it reads like status

File: integration-tests/bats/status.bats
Comment (macneale4) [inline integration-tests/bats/status.bats:462]: can you use identifiers which are 3 chars or more? Easier for future maintainers to search/replace. I know we do single char vars all over the place - just trying to raise the bar one change at a time :-)

File: go/cmd/dolt/commands/status.go
Comment (macneale4) [inline go/cmd/dolt/commands/status.go:120]: This is not a useful comment. It's clear that's what happening by the function name

File: go/cmd/dolt/commands/status.go
Comment (macneale4) [inline go/cmd/dolt/commands/status.go:36]: Brief godoc here would help the reader know the purpose of this structure. Not essential, but when combined with the undocuments createPrintData function, I'm forced to infer that there is a stage where you read all the essential data, then follow that up with printing all the data.

I know about this staging because of comments in the Exec method. My style is to flip this on it's head, and put most comments outside of the code on structure and function definitions. This seems to be the pattern followed on the dolt code I've been exposed to

File: go/cmd/dolt/commands/status.go
Comment (macneale4) [inline go/cmd/dolt/commands/status.go:408]: This function is 200 lines long. Possible to break into bite size chunks?

File: go/libraries/doltcore/sqle/dprocedures/init.go
Comment (macneale4) [inline go/libraries/doltcore/sqle/dprocedures/init.go:33]: Is this a public procedure? Do we need to document this somewhere for user consumption?

**Review by macneale4 — CHANGES_REQUESTED** [review]
This was a lot of work! Broad feedback that I want to give you you should factor your code into smaller functions. There is a pattern of long functions with comments  to indicate processing steps. My feedback is to drop the comments, and convert into single use functions with descriptive names. Other than that, I'm happy with how this is coming. Not a trivial command at all.

File: integration-tests/bats/sql-local-remote.bats
Comment (PavelSafronov) [reply integration-tests/bats/sql-local-remote.bats:129]: Adding the comparison between local and remote executions.

File: integration-tests/bats/status.bats
Comment (PavelSafronov) [reply integration-tests/bats/status.bats:462]: Using (slightly) more descriptive identifiers.

File: go/cmd/dolt/commands/status.go
Comment (PavelSafronov) [reply go/cmd/dolt/commands/status.go:408]: Splitting out most of the "gathering" logic into smaller functions.

File: go/libraries/doltcore/sqle/dprocedures/init.go
Comment (PavelSafronov) [reply go/libraries/doltcore/sqle/dprocedures/init.go:33]: Yup, this is a public procedure. We'll need to add documentation for it to https://github.com/dolthub/docs
I'll be opening a PR for docs after this PR is merged.

File: integration-tests/bats/sql-local-remote.bats
Comment (macneale4) [reply integration-tests/bats/sql-local-remote.bats:129]: You can go further and drop all of the line - line validation. Up to you.

Comment (PavelSafronov) [reply integration-tests/bats/sql-local-remote.bats:129]: I think there's value to doing at least one check of the contents, so I'll remove the repeat checks.

File: integration-tests/bats/status.bats
Comment (macneale4) [reply integration-tests/bats/status.bats:462]: Looks like you changed them in sql-local-remote.bats, but not in status.bats. I'm not going to block on this. Do as you see fit.

**Comment by macneale4** [comment]
Related to: https://github.com/dolthub/dolt/issues/3922

**Pull request merged**
